### DISCUSSION TITLE
Is is possible to animate a path?

### DISCUSSION THREAD

**Discussion by Vadoola** [slint-ui/slint#2722] [Q&A] [answered]
I was playing around with an idea for something and was trying to figure out how to animate a path. So far I haven't found a way that isn't very simplistic.

For example: I can have a path move between 2 states easily enough using a ternary operator for the position values of some of the path components. This can been seen [here](https://slint-ui.com/releases/1.0.2/editor/?snippet=import+%7B+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++VerticalBox+%7B%0A++++++++Rectangle+%7B%0A++++++++++++height%3A+50px%3B%0A++++++++++++width%3A+50px%3B%0A++++++++++++%2F%2Fbackground%3A+red%3B%0A++++++++++++property%3Cbool%3E+active%3A+false%3B%0A++++++++++++TouchArea+%7B%0A++++++++++++++++clicked+%3D%3E+%7Bactive+%3D+%21active%7D%0A++++++++++++%7D%0A%0A++++++++++++txt+%3A%3D+Text+%7B%0A++++++++++++%7D%0A%0A++++++++++++Path+%7B%0A++++++++++++++++viewbox-height%3A+parent.height%2F1px%3B%0A++++++++++++++++viewbox-width%3A+parent.width%2F1px%3B%0A++++++++++++++++stroke%3A+red%3B%0A++++++++++++++++stroke-width%3A+2px%3B%0A%0A++++++++++++++++mt+%3A%3D+MoveTo+%7B%0A++++++++++++++++++++x%3A+0%3B%0A++++++++++++++++++++y%3A+active+%3F+0+%3A+25%3B%0A%0A++++++++++++++++++++animate+y+%7B%0A++++++++++++++++++++++++duration%3A+1000ms%3B%0A++++++++++++++++++++%7D%0A++++++++++++++++%7D%0A++++++++++++++++lt+%3A%3D+LineTo+%7B%0A++++++++++++++++++++x%3A+50%3B%0A++++++++++++++++++++y%3A+active+%3F+50+%3A+25%3B%0A++++++++++++++++++++%0A++++++++++++++++++++animate+y+%7B%0A++++++++++++++++++++++++duration%3A+1000ms%3B%0A++++++++++++++++++++%7D%0A++++++++++++++++%7D%0A++++++++++++++++Close+%7B%7D%0A++++++++++++%7D%0A%0A++++++++++++states+%5B%0A++++++++++++++++on+when+active+%3A+%7B%0A++++++++++++++++++++background%3A+%23bbb%3B%0A++++++++++++++++++++txt.text%3A+%22hi%22%3B%0A++++++++++++++++%7D%0A++++++++++++++++%0A++++++++++++++++off+when+%21active+%3A+%7B%0A++++++++++++++++++++background%3A+white%3B%0A++++++++++++++++++++txt.text%3A+%22bye%22%3B%0A++++++++++++++++%7D%0A++++++++++++%5D%0A++++++++%7D%0A++++%7D%0A%7D%0A), but my code to "animate y" compiles, but just seems to be ignored, as the transition to to the new coordinates is instantaneous.  Also if you want more than 2 simple states, doing this with the ternary operator could get a bit unwieldy.

If I try to add states to a portion of the path (in the link below the initial MoveTo), once again, it compiles, but the states just seem to be completely ignored, and nothing changes. This can be seen [here](https://slint-ui.com/releases/1.0.2/editor/?snippet=import+%7B+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++VerticalBox+%7B%0A++++++++Rectangle+%7B%0A++++++++++++height%3A+50px%3B%0A++++++++++++width%3A+50px%3B%0A++++++++++++%2F%2Fbackground%3A+red%3B%0A++++++++++++property%3Cbool%3E+active%3A+false%3B%0A++++++++++++TouchArea+%7B%0A++++++++++++++++clicked+%3D%3E+%7Bactive+%3D+%21active%7D%0A++++++++++++%7D%0A%0A++++++++++++txt+%3A%3D+Text+%7B%0A++++++++++++%7D%0A%0A++++++++++++Path+%7B%0A++++++++++++++++viewbox-height%3A+parent.height%2F1px%3B%0A++++++++++++++++viewbox-width%3A+parent.width%2F1px%3B%0A++++++++++++++++stroke%3A+red%3B%0A++++++++++++++++stroke-width%3A+2px%3B%0A%0A++++++++++++++++mt+%3A%3D+MoveTo+%7B%0A++++++++++++++++++++x%3A+0%3B%0A++++++++++++++++++++y%3A+50%3B%0A%0A++++++++++++++++++++states+%5B%0A++++++++++++++++++++++++mt_on+when+active+%3A+%7B%0A++++++++++++++++++++++++++++y%3A+0%3B%0A++++++++++++++++++++++++%7D%0A%0A++++++++++++++++++++++++mt_off+when+%21active+%3A+%7B%0A++++++++++++++++++++++++++++y%3A+25%3B%0A++++++++++++++++++++++++%7D%0A++++++++++++++++++++%5D%0A++++++++++++++++%7D%0A++++++++++++++++lt+%3A%3D+LineTo+%7B%0A++++++++++++++++++++x%3A+50%3B%0A++++++++++++++++++++y%3A+25%3B%0A++++++++++++++++%7D%0A++++++++++++++++Close+%7B%7D%0A++++++++++++%7D%0A%0A++++++++++++states+%5B%0A++++++++++++++++on+when+active+%3A+%7B%0A++++++++++++++++++++background%3A+%23bbb%3B%0A++++++++++++++++++++txt.text%3A+%22hi%22%3B%0A++++++++++++++++%7D%0A++++++++++++++++%0A++++++++++++++++off+when+%21active+%3A+%7B%0A++++++++++++++++++++background%3A+white%3B%0A++++++++++++++++++++txt.text%3A+%22bye%22%3B%0A++++++++++++++++%7D%0A++++++++++++%5D%0A++++++++%7D%0A++++%7D%0A%7D%0A).

The third thing I tried is adding the state to a parent (in this case a rectangle). Though the states on the rectangle I can modify various things like the color of the rectangle, and the value of a text object, however if I try to modify the position values of the one of the sub items under the path (such as the MoveTo) it won't compile. This can be seen [here](https://slint-ui.com/releases/1.0.2/editor/?snippet=import+%7BVerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++VerticalBox+%7B%0A++++++++Rectangle+%7B%0A++++++++++++height%3A+50px%3B%0A++++++++++++width%3A+50px%3B%0A++++++++++++property%3Cbool%3E+active%3A+false%3B%0A++++++++++++TouchArea+%7B%0A++++++++++++++++clicked+%3D%3E+%7Bactive+%3D+%21active%7D%0A++++++++++++%7D%0A%0A++++++++++++txt+%3A%3D+Text+%7B%0A++++++++++++%7D%0A%0A++++++++++++Path+%7B%0A++++++++++++++++viewbox-height%3A+parent.height%2F1px%3B%0A++++++++++++++++viewbox-width%3A+parent.width%2F1px%3B%0A++++++++++++++++stroke%3A+red%3B%0A++++++++++++++++stroke-width%3A+2px%3B%0A%0A++++++++++++++++mt+%3A%3D+MoveTo+%7B%0A++++++++++++++++++++x%3A+0%3B%0A++++++++++++++++++++y%3A+25%3B%0A++++++++++++++++%7D%0A++++++++++++++++lt+%3A%3D+LineTo+%7B%0A++++++++++++++++++++x%3A+50%3B%0A++++++++++++++++++++y%3A+25%3B%0A++++++++++++++++%7D%0A++++++++++++++++Close+%7B%7D%0A++++++++++++%7D%0A%0A++++++++++++states+%5B%0A++++++++++++++++on+when+active+%3A+%7B%0A++++++++++++++++++++background%3A+%23bbb%3B%0A++++++++++++++++++++txt.text%3A+%22hi%22%3B%0A++++++++++++++++++++%0A++++++++++++++++++++%2F%2FIf+I+uncomment+out+the+below+line+this+won%27t+compile%0A++++++++++++++++++++%2F%2Fmt.y%3A+0%3B%0A++++++++++++++++%7D%0A++++++++++++++++%0A++++++++++++++++off+when+%21active+%3A+%7B%0A++++++++++++++++++++background%3A+white%3B%0A++++++++++++++++++++txt.text%3A+%22bye%22%3B%0A++++++++++++++++++++%0A++++++++++++++++++++%2F%2FIf+I+uncomment+out+the+below+line+this+won%27t+compile%0A++++++++++++++++++++%2F%2Fmt.y%3A+25%3B%0A++++++++++++++++%7D%0A++++++++++++%5D%0A++++++++%7D%0A++++%7D%0A%7D%0A).

Am I just doing something wrong with my understanding of these animations and path transitions, or is this just not a feature that's currently supported?

**Comment by Vadoola** [comment, marked as answer]
I was actually able to get the path animating using States, and properties. I creating a property for the values that need to change, and then animate the property. The positions of the path are bound to the properties and get updated accordingly as the property changes. This can be seen [here](https://slint-ui.com/releases/1.0.2/editor/?snippet=import+%7BVerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++VerticalBox+%7B%0A++++++++Rectangle+%7B%0A++++++++++++property%3Cint%3E+y_val%3A+25%3B%0A++++++++++++animate+y_val+%7B%0A++++++++++++++++duration%3A+500ms%3B%0A++++++++++++%7D%0A%0A++++++++++++height%3A+50px%3B%0A++++++++++++width%3A+50px%3B%0A++++++++++++property%3Cbool%3E+active%3A+false%3B%0A++++++++++++TouchArea+%7B%0A++++++++++++++++clicked+%3D%3E+%7Bactive+%3D+%21active%7D%0A++++++++++++%7D%0A%0A++++++++++++txt+%3A%3D+Text+%7B%0A++++++++++++%7D%0A%0A++++++++++++Path+%7B%0A++++++++++++++++viewbox-height%3A+parent.height%2F1px%3B%0A++++++++++++++++viewbox-width%3A+parent.width%2F1px%3B%0A++++++++++++++++stroke%3A+red%3B%0A++++++++++++++++stroke-width%3A+2px%3B%0A%0A++++++++++++++++mt+%3A%3D+MoveTo+%7B%0A++++++++++++++++++++x%3A+0%3B%0A++++++++++++++++++++y%3A+y_val%3B%0A++++++++++++++++%7D%0A++++++++++++++++lt+%3A%3D+LineTo+%7B%0A++++++++++++++++++++x%3A+50%3B%0A++++++++++++++++++++y%3A+25%3B%0A++++++++++++++++%7D%0A++++++++++++++++Close+%7B%7D%0A++++++++++++%7D%0A%0A++++++++++++states+%5B%0A++++++++++++++++on+when+active+%3A+%7B%0A++++++++++++++++++++background%3A+%23bbb%3B%0A++++++++++++++++++++txt.text%3A+%22hi%22%3B%0A++++++++++++++++++++y_val%3A+0%3B%0A++++++++++++++++%7D%0A++++++++++++++++%0A++++++++++++++++off+when+%21active+%3A+%7B%0A++++++++++++++++++++background%3A+blue%3B%0A++++++++++++++++++++txt.text%3A+%22bye%22%3B%0A++++++++++++++++++++y_val%3A+25%3B%0A++++++++++++++++%7D%0A++++++++++++%5D%0A++++++++%7D%0A++++%7D%0A%7D%0A).

I do wonder if this should be considered a bug though. Obviously the path can be animated, but having to go through a property, feels a bit cumbersome if I'm statically determining the animation, also since some of the other things I tried compile just fine but don't work is confusing.

**Reply by tronical** [reply]
Well spotted, this is a good workaround. I agree though, it would be good if the implementation of the `Path` API allowed for using animations on `x`, `y`, etc. and all the other _properties_. Right now an implementation detail (that they aren't traditional properties) is leaking through. This is important feedback for #1742.
